packages/react/src/jsx/ReactJSXElement.js

```diff
@@ -259,14 +272,157 @@ export function jsx(type, config, maybeKey) {
   );
 }
 
+// While `jsxDEV` should never be called when running in production, we do
+// support `jsx` and `jsxs` when running in development. This supports the case
+// where a third-party dependency ships code that was compiled for production;
+// we want to still provide warnings in development.
+//
+// So these functions are the _dev_ implementations of the _production_
+// API signatures.
+//
+// Since these functions are dev-only, it's ok to add an indirection here. They
+// only exist to provide different versions of `isStaticChildren`. (We shouldn't
+// use this pattern for the prod versions, though, because it will add an call
+// frame.)
+export function jsxProdSignatureRunningInDevWithDynamicChildren(
+  type,
+  config,
+  maybeKey,
+  source,
+  self,
+) {
+  if (__DEV__) {
+    const isStaticChildren = false;
+    return jsxDEV(type, config, maybeKey, isStaticChildren, source, self);
+  }
+}
+
+export function jsxProdSignatureRunningInDevWithStaticChildren(
+  type,
+  config,
+  maybeKey,
+  source,
+  self,
+) {
+  if (__DEV__) {
+    const isStaticChildren = true;
+    return jsxDEV(type, config, maybeKey, isStaticChildren, source, self);
+  }
+}
+
+const didWarnAboutKeySpread = {};
+
 /**
  * https://github.com/reactjs/rfcs/pull/107
  * @param {*} type
  * @param {object} props
  * @param {string} key
  */
-export function jsxDEV(type, config, maybeKey, source, self) {
+export function jsxDEV(type, config, maybeKey, isStaticChildren, source, self) {
```

Comment:
Other than the soon-to-be-gone-self warning, this shouldn't really be used anymore so the relationship should probably be the inverse where `jsxDEV` helper calls `jsx` or `jsxs` internally for backwards compat but doesn't have its own implementation.

Reply:
Yeah that was how I started but since the implementation `jsx` and `jsxs` are identical in dev right now except for one warning, I was calling a shared function anyway and it ended up being about the same thing. But I'll keep this in mind. I think I can refactor this function so that check isn't so deep and we can get rid of the boolean argument. But I didn't want to do any extra refactoring in this initial PR to minimize the risk of an accidental behavior change.